sdlib/dmd/gc.d

```diff
@@ -67,48 +67,45 @@ bool __sd_gc_fetch_alloc_info(void* ptr, void** base, size_t* size,
 	return true;
 }
 
-size_t __sd_gc_get_array_used(void* ptr, PageDescriptor pd) {
-	auto e = pd.extent;
-	if (!e) {
-		return 0;
-	}
-
-	if (pd.isSlab()) {
-		auto si = SlabAllocInfo(pd, ptr);
-		return si.usedCapacity;
-	} else {
-		return e.usedCapacity;
-	}
-}
-
-// TODO: this should only do large blocks, and let the druntime side do slabs.
-bool __sd_gc_set_array_used(void* ptr, PageDescriptor pd, size_t newUsed,
-                            size_t existingUsed) {
+bool __sd_gc_shrink_array_used(void* ptr, size_t newUsed, size_t existingUsed) {
+	auto pd = threadCache.maybeGetPageDescriptor(ptr);
 	auto e = pd.extent;
 	if (!e) {
 		return false;
 	}
 
 	if (pd.isSlab()) {
 		auto si = SlabAllocInfo(pd, ptr);
-		if (existingUsed < size_t.max && existingUsed != si.usedCapacity) {
-			// The existing capacity doesn't match.
+		auto offset = ptr - si.address + 1;
+		if (existingUsed != 0 && !threadCache
+			    .validateCapacity(ptr[0 .. existingUsed + 1], si.address,
+			                      si.usedCapacity)) {
 			return false;
 		}
 
-		return si.setUsedCapacity(newUsed);
-	} else {
-		if (existingUsed < size_t.max && existingUsed != e.usedCapacity) {
-			// The existing capacity doesn't match.
-			return false;
-		}
+		assert(newUsed + offset <= si.usedCapacity);
+		return si.setUsedCapacity(newUsed + offset);
```

Comment:
Which means that offset doesn't represented the offset of the slice in the slot anymore. I would change that, this might be correct, but very confusing.

Reply:
It was more relevant before, when I was using the same calculation for multiple things. Now it's not. So I've changed it.